---
title: "10-CUTandTag-DeepTools"
output: html_document
date: "2023-10-13"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Set up

```{r Load Libraries}

```

```{r Make Directories}
proj_name <- "OCIAML3_scCUTTag"
scratch_dir <- file.path("scratch", "u", "jo16262", proj_name)
jobreport_dir <- file.path("job_reports")
diffbind_dir <- file.path("diffbind")
diffbind_res_dir <- file.path(diffbind_dir, "results")
bam_dir <- file.path("alignment", "bam")
deeptools_dir <- file.path("deeptools")
figure_dir <- file.path("figures")


if(!dir.exists(deeptools_dir)){
  dir.create(deeptools_dir)
}
```

```{r Set Environmental Variables}
Sys.setenv(SCRATCH_DIR = paste0("/", scratch_dir)) #Accessing the scratch directory requires a backslash before the filepath
Sys.setenv(JOBREP_DIR = paste0("/", scratch_dir, "/", jobreport_dir))
Sys.setenv(BAM_DIR = paste0("/", scratch_dir, "/", bam_dir))
Sys.setenv(SAM_DIR = paste0("/", scratch_dir, "/", sam_dir))
Sys.setenv(DIFFBIND_RES_DIR = paste0("/", scratch_dir, "/", diffbind_res_dir))
Sys.setenv(DEEPTOOLS_DIR = paste0("/", scratch_dir, "/", deeptools_dir))
Sys.setenv(FIGURE_DIR = paste0("/", scratch_dir, "/", figure_dir))
```


## Create BigWig Files
Deep Tools uses big wig files and sorted bam files.
```{cat Create BigWig Files, engine.opts=list(file = 'scripts/bigWigCreation.sh')}
#!/bin/bash

#SBATCH -p normal # partition name
#SBATCH --job-name=CUTTag_bigWig # Job name
#SBATCH --ntasks=1 # number of tasks
#SBATCH --time 0-05:00 # time limit (day-hour:min)
#SBATCH --cpus-per-task=12 # number of threads
#SBATCH --mem-per-cpu=9Gb # requested memory
#SBATCH --account=srrao # PI's net ID
#SBATCH -o job_reports/%x.out # File to which standard outpout will be written
#SBATCH -e job_reports/%x.err # File to whcih standard error wil be written
#SBATCH --mail-type=ALL # What email updates to send
#SBATCH --mail-user=jmurray@mcw.edu # Email adress to send to
echo Starting at $(date)
echo Job name: ${SLURM_JOB_NAME}, Job ID: ${SLURM_JOB_ID}
echo I have ${SLURM_CPUS_ON_NODE} CPUs on compute node $(hostname -s)

projPath="/scratch/u/jo16262/OCIAML3_scCUTTag"


bamPath="${projPath}/alignment/bam"
bigWigPath="${projPath}/alignment/bigwig"

mkdir $bigWigPath

cd $bamPath

module load samtools

for i in $(ls *allreps_bowtie2.mapped.rmDup.bam | rev | cut -c 26- | rev | sort | uniq)
do
samtools sort -@ ${SLURM_CPUS_ON_NODE} -o "$i"_bowtie2.mapped.sorted.rmDup.bam "$i"_bowtie2.mapped.rmDup.bam

samtools index -@ ${SLURM_CPUS_ON_NODE} "$i"_bowtie2.mapped.sorted.rmDup.bam

done

module load deeptools

for i in $(ls *allreps_bowtie2.mapped.sorted.rmDup.bam | rev | cut -c 33- | rev | sort | uniq)
do
bamCoverage -b "$i"_bowtie2.mapped.sorted.rmDup.bam -o "$bigWigPath"/"$i"_bowtie2.bw --extendReads 150 --normalizeUsing BPM
done


```

```{bash, engine.opts='-l'}
cd $SCRATCH_DIR

sbatch scripts/bigWigCreation.sh
```


We also want to create subset bigWig files for Rad21 NTC that correspond to sites with differential binding.
```{cat Create Subset BigWig Files, engine.opts=list(file = 'scripts/subsetBigWigCreation.sh')}
#!/bin/bash

#SBATCH -p normal # partition name
#SBATCH --job-name=CUTTag_subsetBigWig # Job name
#SBATCH --ntasks=1 # number of tasks
#SBATCH --time 0-05:00 # time limit (day-hour:min)
#SBATCH --cpus-per-task=12 # number of threads
#SBATCH --mem-per-cpu=9Gb # requested memory
#SBATCH --account=srrao # PI's net ID
#SBATCH -o job_reports/%x.out # File to which standard outpout will be written
#SBATCH -e job_reports/%x.err # File to whcih standard error wil be written
#SBATCH --mail-type=ALL # What email updates to send
#SBATCH --mail-user=jmurray@mcw.edu # Email adress to send to
echo Starting at $(date)
echo Job name: ${SLURM_JOB_NAME}, Job ID: ${SLURM_JOB_ID}
echo I have ${SLURM_CPUS_ON_NODE} CPUs on compute node $(hostname -s)

projPath="/scratch/u/jo16262/OCIAML3_scCUTTag"


bamPath="${projPath}/alignment/bam"
bigWigPath="${projPath}/alignment/bigwig"
diffBindPath="${projPath}/diffbind/results"


cd $bamPath

##We need to subset the Rad21 NTC bam files using the bed files from diffBind. This can be done using bedtools intersect function.

module load bedtools2

bedtools intersect -wa -abam OCIAML3-NTC_Rad21_bulk-allreps_bowtie2.mapped.rmDup.bam -b "$diffBindPath"/rad21.smc3hetDown.diffbind.bed > OCIAML3-NTC_Rad21_bulk-allreps-hetDown_bowtie2.mapped.rmDup.bam

bedtools intersect -wa -abam OCIAML3-NTC_Rad21_bulk-allreps_bowtie2.mapped.rmDup.bam -b "$diffBindPath"/rad21.smc3hetUp.diffbind.bed > OCIAML3-NTC_Rad21_bulk-allreps-hetUp_bowtie2.mapped.rmDup.bam

##Sort and Index using samtools

module load samtools

samtools sort -@ ${SLURM_CPUS_ON_NODE} -o OCIAML3-NTC_Rad21_bulk-allreps-hetDown_bowtie2.mapped.sorted.rmDup.bam OCIAML3-NTC_Rad21_bulk-allreps-hetDown_bowtie2.mapped.rmDup.bam

samtools sort -@ ${SLURM_CPUS_ON_NODE} -o OCIAML3-NTC_Rad21_bulk-allreps-hetUp_bowtie2.mapped.sorted.rmDup.bam OCIAML3-NTC_Rad21_bulk-allreps-hetUp_bowtie2.mapped.rmDup.bam

samtools index -@ ${SLURM_CPUS_ON_NODE} OCIAML3-NTC_Rad21_bulk-allreps-hetDown_bowtie2.mapped.sorted.rmDup.bam

samtools index -@ ${SLURM_CPUS_ON_NODE} OCIAML3-NTC_Rad21_bulk-allreps-hetUp_bowtie2.mapped.sorted.rmDup.bam

## Create BigWig files

module load deeptools

bamCoverage -b OCIAML3-NTC_Rad21_bulk-allreps-hetDown_bowtie2.mapped.sorted.rmDup.bam -o "$bigWigPath"/OCIAML3-NTC_Rad21_bulk-allreps-hetDown_bowtie2.bw --extendReads 150 --normalizeUsing BPM

bamCoverage -b OCIAML3-NTC_Rad21_bulk-allreps-hetUp_bowtie2.mapped.sorted.rmDup.bam -o "$bigWigPath"/OCIAML3-NTC_Rad21_bulk-allreps-hetUp_bowtie2.bw --extendReads 150 --normalizeUsing BPM



```

```{bash, engine.opts='-l'}
cd $SCRATCH_DIR

sbatch scripts/subsetBigWigCreation.sh
```



##Deep Tools Analysis

The bed tools program needs two files to perform analysis. The first file is the bigwig files we created in the script above. The second file needed is a bed file with peak regions.
```{cat Compute deepTools Matrices, engine.opts=list(file = 'scripts/deepToolMatrix.sh')}
#!/bin/bash

#SBATCH -p normal # partition name
#SBATCH --job-name=CUTTag_deepToolMatrix # Job name
#SBATCH --ntasks=1 # number of tasks
#SBATCH --time 0-10:00 # time limit (day-hour:min)
#SBATCH --cpus-per-task=12 # number of threads
#SBATCH --mem-per-cpu=9Gb # requested memory
#SBATCH --account=srrao # PI's net ID
#SBATCH -o job_reports/%x.out # File to which standard outpout will be written
#SBATCH -e job_reports/%x.err # File to whcih standard error wil be written
#SBATCH --mail-type=ALL # What email updates to send
#SBATCH --mail-user=jmurray@mcw.edu # Email adress to send to
echo Starting at $(date)
echo Job name: ${SLURM_JOB_NAME}, Job ID: ${SLURM_JOB_ID}
echo I have ${SLURM_CPUS_ON_NODE} CPUs on compute node $(hostname -s)

projPath="/scratch/u/jo16262/OCIAML3_scCUTTag"


bigWigPath="${projPath}/alignment/bigwig"
peakPath="${projPath}/peakCalling/consensus"
deepPath="${projPath}/deeptools"
annotationPath="${projPath}/ucsc_tables"

cd $peakPath

module load deeptools

#for i in $(ls *-macs2_peak_q0.05.control.rmDup_peaks.narrowPeak | rev | cut -c 49- | rev | sort | uniq)
#do

# This will create a matrix for a 6 kb region around consensus peaks. We are using the peak summits from MACS2 consensus peaks.
#computeMatrix reference-point --referencePoint center \
#-b 3000 -a 3000 \
#-R "$peakPath"/"$i"-macs2_peak_q0.05.control.rmDup_summits.bed \
#--skipZeros \
#-S "$bigWigPath"/"$i"_bowtie2.bw -p ${SLURM_CPUS_ON_NODE} \
#-o "$deepPath"/"$i"_consensusPeaks.matrix.gz \
#--outFileSortedRegions "$deepPath"/"$i"_consensusPeaks.regions.bed

## -R is the regions we want to make a matrix for (aka peaks)
## -S are the subject files (bigWig files)

#done

cd $bigWigPath

module load deeptools

#for i in $(ls *Rad21*allreps*het*_bowtie2.bw | rev | cut -c 12- | rev | awk -v FS='_' '{print $2"_"$3}' | sort | uniq)
for i in $(ls *bulk-allreps-hetDown_bowtie2.bw | rev | cut -c 16- | rev | uniq)

do
# This will create a matrix over transcription sites
computeMatrix scale-regions \
-R "$annotationPath"/hg38_refseq_genes.bed \
--skipZeros \
-S "$bigWigPath"/"$i"*_bowtie2.bw -p ${SLURM_CPUS_ON_NODE} \
--beforeRegionStartLength 3000 \
--regionBodyLength 5000 \
--afterRegionStartLength 3000 \
-o "$deepPath"/"$i"_diffbind_tss.matrix.gz \
--outFileSortedRegions "$deepPath"/"$i"_diffbind_tss.regions.bed

# This will create a matrix for proximal non-promoter & non-ctcf cis-regulatory elements (enhancers) from ENCODE
computeMatrix reference-point \
-b 3000 -a 3000 \
-R "$annotationPath"/hg38_ccres_noPLSnoCTCF.bed \
--skipZeros \
-S "$bigWigPath"/"$i"*_bowtie2.bw -p ${SLURM_CPUS_ON_NODE} \
-o "$deepPath"/"$i"_diffbind_enhancers.matrix.gz \
--outFileSortedRegions "$deepPath"/"$i"_diffbind_enhancers.regions.bed

# This will create a matrix for CTCF-sites (also from ENCODE)
computeMatrix reference-point \
-b 3000 -a 3000 \
-R "$annotationPath"/hg38_ccres_CTCF.bed \
--skipZeros \
-S "$bigWigPath"/"$i"*_bowtie2.bw -p ${SLURM_CPUS_ON_NODE} \
-o "$deepPath"/"$i"_diffbind_ctcf.matrix.gz \
--outFileSortedRegions "$deepPath"/"$i"_diffbind_ctcf.regions.bed

# This will create a matrix for promoter sites (data also from ENCODE cCREs)
computeMatrix reference-point \
-b 3000 -a 3000 \
-R "$annotationPath"/hg38_ccres_promoters.bed \
--skipZeros \
-S "$bigWigPath"/"$i"*_bowtie2.bw -p ${SLURM_CPUS_ON_NODE} \
-o "$deepPath"/"$i"_diffbind_promoters.matrix.gz \
--outFileSortedRegions "$deepPath"/"$i"_diffbind_promoters.regions.bed

done


```


```{bash, engine.opts='-l'}
cd $SCRATCH_DIR

sbatch scripts/deepToolMatrix.sh
```


##DeepTools Visualization

```{cat Compute deepTools Matrices, engine.opts=list(file = 'scripts/deepToolVisualize.sh')}
#!/bin/bash

#SBATCH -p normal # partition name
#SBATCH --job-name=CUTTag_deepToolVisualize # Job name
#SBATCH --ntasks=1 # number of tasks
#SBATCH --time 0-05:00 # time limit (day-hour:min)
#SBATCH --cpus-per-task=12 # number of threads
#SBATCH --mem-per-cpu=9Gb # requested memory
#SBATCH --account=srrao # PI's net ID
#SBATCH -o job_reports/%x.out # File to which standard outpout will be written
#SBATCH -e job_reports/%x.err # File to whcih standard error wil be written
#SBATCH --mail-type=ALL # What email updates to send
#SBATCH --mail-user=jmurray@mcw.edu # Email adress to send to
echo Starting at $(date)
echo Job name: ${SLURM_JOB_NAME}, Job ID: ${SLURM_JOB_ID}
echo I have ${SLURM_CPUS_ON_NODE} CPUs on compute node $(hostname -s)

projPath="/scratch/u/jo16262/OCIAML3_scCUTTag"


bigWigPath="${projPath}/alignment/bigwig"
peakPath="${projPath}/peakCalling/consensus"
deepPath="${projPath}/deeptools"
diffBindPath="${projPath}/diffbind/results"
figurePath="${projPath}/figures"

cd $deepPath

module load deeptools

#Create profile plots for cis-regulatory elements
plotProfile -m Rad21_bulk-allreps_ctcf.matrix.gz \
-out "$figurePath"/ociaml3_rad21_at_ENCODEctcf_profilePlot.png \
--regionsLabel "" \
--perGroup \
--plotTitle "Rad21 Signal at ENCODE CTCF Binding Sites" \
--refPointLabel "CTCF CREs"

#Create profile plots for transcription sites

#Create profile plots for peaks
plotProfile -m OCIAML3-NTC_Rad21_bulk-allreps_consensusPeaks.matrix.gz \
-out "$figurePath"/ociaml3-NTC_rad21_peaks_profilePlot.png \
--regionsLabel "" \
--perGroup \
--plotTitle "Rad21 NTC OCI-AML3" \
--refPointLabel "Peak Center"

plotProfile -m OCIAML3-SMC3het_Rad21_bulk-allreps_consensusPeaks.matrix.gz \
-out "$figurePath"/ociaml3-SMC3het_rad21_peaks_profilePlot.png \
--regionsLabel "" \
--perGroup \
--plotTitle "Rad21 SMC3het OCI-AML3" \
--refPointLabel "Peak Center"


```


```{bash, engine.opts='-l'}
cd $SCRATCH_DIR

sbatch scripts/deepToolVisualize.sh
```

```{bash, engine.opts='-l'}
cd $DEEPTOOLS_DIR

module load deeptools

plotProfile -m Rad21_bulk-allreps_tss.matrix.gz \
-out "$FIGURE_DIR"/ociaml3_rad21_at_refseqGenes_profilePlot.pdf \
--regionsLabel "" \
--perGroup \
--plotTitle "Rad21 Signal at RefSeq Genes" \
--plotHeight 10 \
--plotWidth 15 \
--samplesLabel "Rad21-NTC:SMC3Het Down" "Rad21-NTC:SMC3het Up" "Rad21-NTC:All" "Rad21-SMC3het" \
--legendLocation upper-right \
--colors '#073763' '#366785' '#5b5b5b' '#c85200'

```






